### ISSUE TITLE
How to calculate the profit / loss along with the fee for each buy / sell market order pair

### ISSUE THREAD

**Issue by wwalterr** [ccxt/ccxt#14754]
I'm using Python, CCXT and Kucoin to create buy / sell market orders and I want to calculate how much each order pair generates in profit / loss, i.e sell - buy.

After reading the order structure [documentation](https://docs.ccxt.com/en/latest/manual.html#order-structure) and the issues [3101](https://github.com/ccxt/ccxt/issues/3101) and [6996](https://github.com/ccxt/ccxt/issues/6996), I'm a little bit confused on how to get the buy and sell price along with the fee since there is the amount, price, filled, cost (filled * price) and the fee, but all of these exchanging between "quote currency" and "base currency" - I assume the documentation is referring to [this](https://github.com/ccxt/ccxt/issues/2707#issuecomment-385580495).

An buy / sell market order pair example:

```python
# Buy order

{
    'symbol': 'BTC/USDT',
    'type': 'market',
    'timeInForce': 'GTC',
    'postOnly': False,
    'side': 'buy',
    'amount': 0.0004,
    'price': 21089.9,
    'stopPrice': 0.0,
    'cost': 8.43596,
    'filled': 0.0004,
    'remaining': 0.0,
    'timestamp': 1660955796571,
    'datetime': '2022-08-20T00:36:36.571Z',
    'fee': {'currency': 'USDT', 'cost': 0.00843596},
    'status': 'closed',
    'lastTradeTimestamp': None,
    'average': 21089.9,
    'trades': [],
    'fees': [{'currency': 'USDT', 'cost': 0.00843596}]
}
```

```python
# Sell order

{
    'symbol': 'BTC/USDT',
    'type': 'market',
    'timeInForce': 'GTC',
    'postOnly': False,
    'side': 'sell',
    'amount': 0.0004,
    'price': 21110.3,
    'stopPrice': 0.0,
    'cost': 8.44412,
    'filled': 0.0004,
    'remaining': 0.0,
    'timestamp': 1660956003764,
    'datetime': '2022-08-20T00:40:03.764Z',
    'fee': {'currency': 'USDT', 'cost': 0.00844412},
    'status': 'closed',
    'lastTradeTimestamp': None,
    'average': 21110.3,
    'trades': [],
    'fees': [{'currency': 'USDT', 'cost': 0.00844412}]
}
```

P.S No details on how to actually calculate are provided on the issues and I want to make sure I'm not calculating it wrong and loosing money if I run my bot 24 / 7

**Comment by pier0074** [comment]
In general : 
Gross profit = Revenue - Cost
Net Profit = Gross profit - Fees

In this case : 
Revenue = Sell cost = 8.44412 which is also equal to Filled * Price = 0.0004 * 21110.3 = 8.44412
Cost = Buy cost = 8.43596 which is also equal to Filled * Price = 0.0004 * 21089.9 = 8.43596
Gross profit  = 8.44412 - 8.43596 = 0.00816
Fees = Sell Fees + Buy Fees = Sell fee.cost + Buy fee.cost = 0.00844412 + 0.00843596 = 0.01688008
Net Profit = 0.00816 -  0.01688008 = -0.00872008

In conclusion, excluding the fees you are profitable on this trade, but all together, you are actually losing money.

**Comment by carlosmiei** [comment]
Hello @pier0074 

I think your math is on point. As a small optimization, I would suggest using the cost field to do it so that you don't need to calculate it manually

**Comment by pier0074** [comment]
Hello @carlosmiei 

This is what I firstly used.
When I mentioned "which is also equal to", I was just pointing out where the cost field comes from.

**Comment by wwalterr** [comment]
Thanks a lot for taking the time and break the calculations on its terms.

I'll change change the exchange, review the cumulative return range and develop the bot in a fee [friendly](https://twitter.com/Solana_Mates/status/1382692005659168782/photo/1) network.

> Some exchanges supports the method `fetch_trading_fees(symbol, params = {})` and the exchange trading tiers can be found in the exchange [code](https://github.com/ccxt/ccxt/blob/master/python/ccxt/kucoin.py#L368)

Code:

```python
def gross_profit(buy_order: dict, sell_order: dict) -> float:
    revenue = sell_order.get('cost')
    
    cost = buy_order.get('cost')
    
    return revenue - cost
```

```python
def net_profit(buy_order: dict, sell_order: dict) -> float:
    fees = buy_order.get('fee').get('cost') + sell_order.get('fee').get('cost')
    
    return gross_profit(buy_order, sell_order) - fees
```

Example:

```python
buy_order = await exchange.fetch_order('ID')

sell_order = await exchange.fetch_order('ID')

gross_profit(buy_order, sell_order)

net_profit(buy_order, sell_order)
```
